### DISCUSSION TITLE
Is it possible to stream local data to the player after create() is called?

### DISCUSSION THREAD

**Discussion by boris-42** [asciinema/discussions#214] [Q&A] [answered]
# Use case
Playing active session, when initially only part of cast is available, as time passes new parts of cast are available. 

# My understanding at the moment 
This will require to write new driver, similar to ws, which is going to directly push new data to buffer when it is available?

If this is true, should I contribute this driver?

**Comment by ku1ik** [comment, marked as answer]
Good thinking, your idea to implement a custom driver seems exactly what I'd suggest in this case.

I'm not sure if I'd be interested in getting it bundled with the player though. What you describe is a hybrid between finite recording driver and potentially infinite websocket driver.

I suggest just implementing it outside of the player. You can do that by creating a function which returns a driver object with at least start method, and passing that function as first arg to `AsciinemaPlayer.create`. See test drivers for examples of minimal drivers. And you don't need to use buffer like built-in websocket driver does, you can just call `feed` directly.

Note to myself: I need to finally make drivers official, document this concept properly.

**Reply by boris-42** [reply]
@sickill Thank you for quick reply!

I believe I am not the only one who might need this.
Maybe we can put it in examples/docs for drivers.
It won't be in a bundle but If someone needs it, they can just copy it.

**Reply by ku1ik** [reply]
Yeah, we could have a `contrib` directory for user-contributed drivers, parsers and other snippets.

**Comment by ku1ik** [comment]
Here are few examples of implementing a custom driver and passing it as `src` arg to `AsciinemaPlayer.create`.

Simplest driver is a function which returns a function. Returned function is called when user starts playback. It may return another function which is used for cleanup when player is removed from DOM. 

```javascript
function src1({ feed }) {
  return () => {
    const intervalId = setInterval(() => feed('x'), 100);

    return () => clearInterval(intervalId);
  }
};

AsciinemaPlayer.create(src1, document.getElementById('demo'));
```

The above can only be started, doesn't provide option to pause or seek.

To implement more functionality, like pause/resume, seek etc a driver function must return an object with various callbacks implemented.

Driver below returns cols/rows (can be done when it's known upfront), `play` and `stop` callbacks. This is still very minimal and pretty much equivalent to the driver above, but allows for implementing more callbacks.

```javascript
function src2({ feed }) {
  let intervalId;

  return {
    cols: 16,
    rows: 8,

    play() {
      intervalId = setInterval(() => feed('x'), 100);
      return true;
    },

    stop() {
      clearInterval(intervalId);
    },
  }
};
```

Below is an extended version of the above one, which has `init` callback which can be used to initialize the driver, e.g. fetch and parse data. This example also shows how you can return terminal size from `init`, e.g. when it's not known upfront but you need to extract it from fetched file.

```javascript
function src3({ feed }) {
  let intervalId;

  return {
    init() {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve({ cols: 16, rows: 8 });
        }, 1500);
      });
    },

    play() {
      intervalId = setInterval(() => feed('x'), 100);
      return true;
    },

    stop() {
      clearInterval(intervalId);
    },
  }
};
```

See recording or websocket drivers for other callbacks like `pause` or `seek`.

I hope this gets you going!

**Comment by ku1ik** [comment]
Another thought about your use case:

If you don't require seeking and pause you can use existing websocket driver. Thanks to its buffer it supports "pre-buffering".

For example, given following "init" cast file:

```
# init.cast
{ "version": 2, "width": 80, "height": 24 }
[0.0, "o", "x"]
[10.0, "o", "y"]
[20.0, "o", "z"]
```

It's possible to feed it into the player, and then later send more data. I tested this with [websocat](https://github.com/vi/websocat):

```bash
mkfifo live.pipe
websocat -v --text --no-close ws-l:127.0.0.1:9002 reuse-raw:autoreconnect:readfile:live.pipe
```

Connected to it in JS like this:

```javascript
AsciinemaPlayer.create('ws://127.0.0.1:9002');
```

Then in another shell:

```bash
cat init.cast >>live.pipe
echo '[21.0, "o", "a"]' >>live.pipe
echo '[22.0, "o", "b"]' >> live.pipe
echo '[23.0, "o", "c"]' >>live.pipe
echo '[24.0, "o", "d"]' >> live.pipe
```

Websocket driver honours timestamp of each event, allowing future events to be sent upfront, and displayed later.
